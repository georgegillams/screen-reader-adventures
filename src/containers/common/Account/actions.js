import { createRoutine } from 'redux-saga-routines';

import { LOGOUT, REQUEST_VERIFICATION_EMAIL } from './constants';

export const logout = createRoutine(LOGOUT);
export const requestVerificationEmail = createRoutine(REQUEST_VERIFICATION_EMAIL);
