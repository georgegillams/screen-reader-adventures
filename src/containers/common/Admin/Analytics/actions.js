import { createRoutine } from 'redux-saga-routines';

import { LOAD } from './constants';

export const load = createRoutine(LOAD);
