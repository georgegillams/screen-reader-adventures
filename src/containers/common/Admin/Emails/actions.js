import { createRoutine } from 'redux-saga-routines';

import { LOAD, RESEND } from './constants';

export const load = createRoutine(LOAD);
export const resend = createRoutine(RESEND);
