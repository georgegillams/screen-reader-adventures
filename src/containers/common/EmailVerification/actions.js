import { createRoutine } from 'redux-saga-routines';

import { VERIFY } from './constants';

export const verify = createRoutine(VERIFY);
