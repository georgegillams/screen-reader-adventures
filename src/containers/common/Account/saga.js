import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { logout, requestVerificationEmail } from './actions';
import { setUser } from 'containers/common/Authenticator/actions';

export function* doLogout() {
  const requestURL = apiStructure.logout.fullPath;

  try {
    yield put(logout.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
    });

    if (result.error) {
      yield put(logout.failure(result));
    } else {
      yield put(logout.success(result));
      yield put(setUser.trigger(null));
    }
  } catch (err) {
    yield put(logout.failure(err));
  } finally {
    yield put(logout.fulfill());
  }
}

export function* doRequestVerificationEmail() {
  const requestURL = apiStructure.requestVerificationEmail.fullPath;

  try {
    yield put(requestVerificationEmail.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
    });

    if (result.error) {
      yield put(requestVerificationEmail.failure(result));
    } else {
      yield put(requestVerificationEmail.success(result));
    }
  } catch (err) {
    yield put(requestVerificationEmail.failure(err));
  } finally {
    yield put(requestVerificationEmail.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(logout.TRIGGER, doLogout);
  yield takeLatest(requestVerificationEmail.TRIGGER, doRequestVerificationEmail);
}
