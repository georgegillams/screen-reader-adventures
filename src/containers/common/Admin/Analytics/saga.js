import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { load } from './actions';

export function* doLoad() {
  const requestURL = apiStructure.loadAnalyticsSummary.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.analytics));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
}
