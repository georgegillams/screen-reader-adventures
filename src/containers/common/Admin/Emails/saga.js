import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { load, resend } from './actions';
import { selectState } from './selectors';

export function* doLoad() {
  const requestURL = apiStructure.loadEmails.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.emails));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export function* doResend() {
  const currentState = yield select(selectState());
  const { emailToResend } = currentState;
  const requestURL = apiStructure.resendEmail.fullPath;

  try {
    yield put(resend.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(emailToResend),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(resend.failure(result));
    } else {
      yield put(resend.success(result.emails));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(resend.failure(err));
  } finally {
    yield put(resend.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
  yield takeLatest(resend.TRIGGER, doResend);
}
