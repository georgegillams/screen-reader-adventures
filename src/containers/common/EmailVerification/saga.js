import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { selectState } from './selectors';
import { verify } from './actions';
import { loadAuth } from 'containers/common/Authenticator/actions';

export function* doVerify() {
  const currentState = yield select(selectState());
  const { token } = currentState;
  const requestURL = apiStructure.verifyEmail.fullPath;

  try {
    yield put(verify.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify({ verificationKey: token }),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(verify.failure(result));
    } else {
      yield put(verify.success(result));
      yield put(loadAuth.trigger());
    }
  } catch (err) {
    yield put(verify.failure(err));
  } finally {
    yield put(verify.fulfill());
  }
}
export default function* saga() {
  yield takeLatest(verify.TRIGGER, doVerify);
}
