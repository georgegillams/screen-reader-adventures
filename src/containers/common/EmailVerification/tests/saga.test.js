import { put, takeLatest } from 'redux-saga/effects';

import { verify } from '../actions';
import { loadAuth } from 'containers/common/Authenticator/actions';

import saga, { doVerify } from '../saga';

describe('EmailVerification saga', () => {
  let mainSaga;

  beforeEach(() => {
    mainSaga = saga();
  });

  it('Should verify on verify TRIGGER', () => {
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(verify.TRIGGER, doVerify));
  });

  describe('account actions', () => {
    let verifyGenerator;

    const response = {
      success: 'you are logged out',
      status: 200,
    };

    beforeEach(() => {
      verifyGenerator = doVerify();

      const selectDescriptor = verifyGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call verify.success on successful API call', () => {
      verifyGenerator.next({ token: 'asdfghjkl' });
      verifyGenerator.next();
      const putSuccess = verifyGenerator.next(response).value;
      verifyGenerator.next();

      expect(putSuccess).toEqual(put(verify.success(response)));
    });

    it('Should call loadAuth.trigger on successful API call', () => {
      verifyGenerator.next({ token: 'asdfghjkl' });
      verifyGenerator.next();
      verifyGenerator.next(response);
      const putSuccess = verifyGenerator.next().value;
      verifyGenerator.next();

      expect(putSuccess).toEqual(put(loadAuth.trigger()));
    });

    it('Should call verify.failure if API returns error', () => {
      verifyGenerator.next({ token: 'asdfghjkl' });
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      verifyGenerator.next();
      const putSuccess = verifyGenerator.next(response).value;
      verifyGenerator.next();

      expect(putSuccess).toEqual(put(verify.failure(response)));
    });

    it('Should call verify.failure if an exception occurs', () => {
      verifyGenerator.next({ token: 'asdfghjkl' });
      const response = new Error('Some error');
      const putFailure = verifyGenerator.throw(response).value;

      expect(putFailure).toEqual(put(verify.failure(response)));
    });
  });
});
