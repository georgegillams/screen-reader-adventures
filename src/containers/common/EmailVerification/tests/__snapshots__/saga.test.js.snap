// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailVerification saga account actions Should call loadAuth.trigger on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`EmailVerification saga account actions Should call verify.failure if API returns error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`EmailVerification saga account actions Should call verify.failure if an exception occurs 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`EmailVerification saga account actions Should call verify.success on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
